libsam.c

```diff
@@ -839,6 +842,178 @@ int sam_put_user_grpids(struct hive *hdesc, int rid, struct keyval *val)
 
 /********* GROUP / USER MANIPULATION ROUTINES **************/
 
+/* Add SID to a group
+ * SID = any SID
+ * grp = group ID
+ * return true if success
+ */
+
+int sam_add_sid_to_grp(struct hive *hdesc, struct sid_binary * sid, int grp)
+{
+    struct sid_array *sarray, *narray;
+    struct sid_binary *usid = sid;
+    int members, newmembers;
+    char *str;
+    int o, n, hit, c;
+
+    if (!sid || !grp || (hdesc->type !=HTYPE_SAM) ) return(0);
+
+    if (gverbose)
+    {
+        str = sam_sid_to_string(usid);
+        printf("sam_add_sid_to_grp: user SID is <%s>\n", str);
+        free(str);
+    }
+
+    /* Just add SID to group, SID without RID situation like AD users
+     */
+
+    members = sam_get_grp_members_sid(hdesc, grp, &sarray);
+
+    if (!sarray) {
+      printf("sam_add_sid_to_grp: group # %x not found!\n",grp);
+      return(0);
+    }
+    if (gverbose)
+    {
+        printf("add_user_to_grp: grp memberlist BEFORE:\n");
+
+        for (o = 0; sarray[o].sidptr; o++)
+        {
+            str = sam_sid_to_string(sarray[o].sidptr);
+            printf("  Member # %d = <%s>\n", o, str);
+            FREE(str);
+            }
+    }
+
+    newmembers = members + 1;
+    ALLOC(narray, sizeof(struct sid_array) * (newmembers + 2), 1);  /* Add one entry size */
+
+    if (gverbose) printf("members = %d\n", members);
+
+    hit = 0;
+    for (o = 0, n = 0; o <= members; o++, n++) {
+      c = sam_sid_cmp(sarray[o].sidptr, usid);     /* Compare slot with new SID */
+      if (gverbose) printf("sam_sid_cmp returns %d\n",c);
+      if (c == 0) {
+        newmembers--;                   /* Already there, don't change anything */
+        hit = 1;
+      }
+      if (!hit && ((c > 0) || !sarray[o].sidptr)) {              /* Next is higher, insert new SID */
+        if (gverbose) printf("  -- add\n");
+        narray[n].len = usid->sections * 4 + 8;     /* Hmm */
+        narray[n].sidptr = usid;
+        n++;
+        hit = 1;
+      }
+      narray[n].len = sarray[o].len;
+      narray[n].sidptr = sarray[o].sidptr;
+    }
+
+    if (gverbose)
+    {
+        printf("sam_add_sid_to_grp: grp memberlist AFTER:\n");
+
+        for (o = 0; narray[o].sidptr; o++)
+        {
+            str = sam_sid_to_string(narray[o].sidptr);
+            printf("  Member # %u = <%s>\n", o, str);
+           FREE(str);
+        }
+    }
```

Comment:
gverbose should only be used to turn on or off the printf statement.
Else if gverbose is false then your code does not do anything useful.